src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.InsertMapperHelper.cs

```diff
@@ -0,0 +1,349 @@
+        /// <param name="insertion">The snippet to insert.</param>
+        /// <param name="target"></param>
+        /// <param name="adjustedNodeToMap"></param>
+        /// <returns></returns>
+        protected override TextSpan? GetInsertSpan(SyntaxNode documentSyntax, NodeToMap insertion, DocumentSpan target, out SyntaxNode? adjustedNodeToMap)
+        {
+            adjustedNodeToMap = null;
+            int insertionPoint;
+            if (insertion.Scope is not Scope.None)
+            {
+                if (TryGetScopedInsertionPoint(documentSyntax, insertion.Scope, out insertionPoint))
+                {
+                    return new TextSpan(insertionPoint, 0);
+                }
+            }
+
+            // If there's an specific focus area, or caret provided, we should try to insert as close as possible.
+            // As long as the focused area is not empty.
+            if (TryGetFocusedInsertionPoint(target.SourceSpan, documentSyntax, insertion, out insertionPoint))
+            {
+                return new TextSpan(insertionPoint, 0);
+            }
+
+            // Fallback: Attempt to infer the insertion point without a caret or line.
+            // This will attempt to get a default insertion point for the insert node within the
+            // current document.
+            if (TryGetDefaultInsertionPoint(documentSyntax, out insertionPoint))
+            {
+                return new TextSpan(insertionPoint, 0);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Tries to get the insertion point for the default case.
+        /// There's many heuristics here that are applied like:
+        /// If what is intended to be inserted is a method, or a class, then we find the right
+        /// place in the document for it. If currently the document is classless, we also take that
+        /// into consideration.
+        /// </summary>
+        /// <param name="node">The node to get insertion point for.</param>
+        /// <param name="insertionPoint">The computed insertion point.</param>
+        /// <returns><c>true</c> if an insertion point was successfully computed,
+        /// <c>false</c> otherwise.</returns>
+        private static bool TryGetDefaultInsertionPoint(SyntaxNode node, out int insertionPoint)
+        {
+            var firstClass = node.DescendantNodesAndSelf()
+                .OfType<ClassDeclarationSyntax>()
+                .FirstOrDefault();
+            if (firstClass is null)
+            {
+                // Get classless insertion.
+                return TryGetClasslessInsertionPoint(node, out insertionPoint);
+            }
+
+            // If last property is not null, get the span end as insertion point.
+            var lastProperty = firstClass.DescendantNodes()
+                .OfType<MemberDeclarationSyntax>()
+                .Where(NodeToMap.IsSimpleNode)
+                .LastOrDefault();
+            if (lastProperty is not null)
```

Comment:
what is being inserted, and why is "after the last proeprty" the right place to put something.

note: we already have logic to determine where to insert code depending on what it's going into and what the node is that we're inserting.  This logic is in the CodeGenerationService, and it's what many features already use.